cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorStrategy.java

```diff
@@ -1,55 +1,61 @@
 package com.sequenceiq.cloudbreak.cloud.aws.service.subnetselector;
 
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.Collectors;
 
-import javax.ws.rs.BadRequestException;
-
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
+import com.sequenceiq.cloudbreak.cloud.model.SubnetSelectionResult;
 
 public abstract class SubnetSelectorStrategy {
 
+    static final String NOT_ENOUGH_AZ = "Acceptable subnets are in %d different AZs, but subnets in %d different AZs required.";
+
     private static final Logger LOGGER = LoggerFactory.getLogger(SubnetSelectorStrategy.class);
 
-    public List<CloudSubnet> select(List<CloudSubnet> subnetMetas) {
+    public SubnetSelectionResult select(List<CloudSubnet> subnetMetas) {
         LOGGER.debug("Subnet selection with strategy '{}'", getType());
-        quickValidate(subnetMetas);
-        List<CloudSubnet> selectedNetworks = selectInternal(subnetMetas);
-        LOGGER.debug("Selected subnets: {}", String.join(", ", selectedNetworks.stream().toString()));
-        return selectedNetworks;
-    }
-
-    private void quickValidate(List<CloudSubnet> subnetMetas) {
-        if (subnetMetas == null || subnetMetas.isEmpty()) {
-            error("There are no subnets in this network.");
-        }
-        if (subnetMetas.size() < getMinimumNumberOfSubnets()) {
-            error(String.format("There are not enough subnets in this network, found: %d, expected: %d.", subnetMetas.size(), getMinimumNumberOfSubnets()));
+        Optional<String> errorMessage = quickValidate(subnetMetas);
+        if (errorMessage.isPresent()) {
+            return new SubnetSelectionResult(errorMessage.get());
         }
+        SubnetSelectionResult selectionResult = selectInternal(subnetMetas);
+        logResult(selectionResult);
+        return selectionResult;
```

Comment:
`else` is missing

Reply:
What do you mean by else? You mean, fallback strategy? That you will find at SubnetIdProvider or at SubnetChooser, as they do different things on fallback (does this answer your question?)

Reply:
you have an if, and this is the else part. makes the code easier to read if you don't have hidden returns

Reply:
I am not sure I agree. I think this is opinion based, as some like single returns, whereas others prefer early returns. I personally think there is value in early returns as it avoids nesting and gives a clear flow to the code.